File: src/main/java/alfheim/client/render/entity/RenderEntityGravityTrap.java
package alfheim.client.render.entity;

import static org.lwjgl.opengl.GL11.*;

import java.util.Random;

import alexsocol.asjlib.render.RenderPostShaders;
import alexsocol.asjlib.render.ShadedObject;
import alfheim.api.ModInfo;
import alfheim.api.lib.LibResourceLocations;
import alfheim.api.lib.LibShaderIDs;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.entity.Render;
import net.minecraft.entity.Entity;
import net.minecraft.util.ResourceLocation;

public class RenderEntityGravityTrap extends Render {

	Random rand = new Random();
	
	public static ShadedObject so = new ShadedObject(LibShaderIDs.idGravity, RenderPostShaders.getNextAvailableRenderObjectMaterialID(), LibResourceLocations.gravity) {

		@Override
		public void preRender() {
			glEnable(GL_BLEND);
			glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);

			glDisable(GL_CULL_FACE);
			glShadeModel(GL_SMOOTH);
			
			// glColor4d(1, 1, 1, chargeMul); 
		}

		@Override
		public void drawMesh() {
			Tessellator tes = Tessellator.instance;
			tes.startDrawingQuads();
			tes.addVertexWithUV(-1, 0, -1, 0, 0);
			tes.addVertexWithUV(-1, 0, 1, 0, 1);
			tes.addVertexWithUV(1, 0, 1, 1, 1);
			tes.addVertexWithUV(1, 0, -1, 1, 0);
			tes.draw();
		}
		
		@Override
		public void postRender() {
			glShadeModel(GL_FLAT);
		}
		
	};
	
	public RenderEntityGravityTrap() {
		super();
		shadowSize = 0.0F;
		RenderPostShaders.registerShadedObject(so);
	}
	
	@Override
	protected ResourceLocation getEntityTexture(Entity p_110775_1_) {
		return null;
	}
	
	@Override
	public void doRender(Entity e, double x, double y, double z, float yaw, float partialTick) {
		glPushMatrix();
		glTranslated(x, y + 0.01, z);

		float live = e.ticksExisted;
		float charge = Math.min(20F, live + partialTick);
		float chargeMul = charge / 20F;
		
		rand.setSeed(e.getUniqueID().getMostSignificantBits());

		float s = chargeMul;
		s += Math.min(1F, (live + partialTick) * 0.2F);
		s *= 2F;
		glScalef(s, s, s);
		
		glRotatef(charge * 9F + (e.ticksExisted + partialTick) * 0.5F + rand.nextFloat() * 360F, 0F, 1F, 0F);

		so.addTranslation();

		glPopMatrix();
	}
	
	
}

File: src/main/java/alfheim/client/render/entity/RenderEntityGravityTrap.kt
package alfheim.client.render.entity

import alexsocol.asjlib.glScaled
import alexsocol.asjlib.render.*
import alfheim.api.lib.*
import alfheim.common.spell.tech.SpellGravityTrap
import net.minecraft.client.renderer.Tessellator
import net.minecraft.client.renderer.entity.Render
import net.minecraft.entity.Entity
import org.lwjgl.opengl.GL11.*
import java.util.*
import kotlin.math.min

object RenderEntityGravityTrap: Render() {
	
	internal val rand = Random()
	
	init {
		shadowSize = 0f
	}
	
	override fun getEntityTexture(p_110775_1_: Entity) = null
	
	override fun doRender(e: Entity, x: Double, y: Double, z: Double, yaw: Float, partialTick: Float) {
		glPushMatrix()
		glTranslated(x, y + 0.01, z)
		
		val charge = min(20f, e.ticksExisted + partialTick)
		
		rand.setSeed(e.uniqueID.mostSignificantBits)
		
		val s = (charge / 20 + min(1f, (e.ticksExisted + partialTick) * 0.2f)) / 2 * SpellGravityTrap.radius
		glScaled(s)
		
		glRotatef(charge * 9f + (e.ticksExisted + partialTick) * 0.5f + rand.nextFloat() * 360f, 0f, 1f, 0f)
		
		so.addTranslation()
		
		glPopMatrix()
	}
	
	val so: ShadedObject = object: ShadedObject(LibShaderIDs.idGravity, RenderPostShaders.nextAvailableRenderObjectMaterialID, LibResourceLocations.gravity) {
		
		override fun preRender() {
			glEnable(GL_BLEND)
			glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA)
			
			glDisable(GL_CULL_FACE)
			glShadeModel(GL_SMOOTH)
			
			// glColor4d(1, 1, 1, chargeMul);
		}
		
		override fun drawMesh() {
			val tes = Tessellator.instance
			tes.startDrawingQuads()
			tes.addVertexWithUV(-1.0, 0.0, -1.0, 0.0, 0.0)
			tes.addVertexWithUV(-1.0, 0.0, 1.0, 0.0, 1.0)
			tes.addVertexWithUV(1.0, 0.0, 1.0, 1.0, 1.0)
			tes.addVertexWithUV(1.0, 0.0, -1.0, 1.0, 0.0)
			tes.draw()
		}
		
		override fun postRender() {
			glShadeModel(GL_FLAT)
		}
	}.also { RenderPostShaders.registerShadedObject(it) }
}